### PR TITLE
drivers/interrupt_controller: stm32: Complete driver factorization

### PR DESCRIPTION
**Pull request by erwango**
Complete code factorization in stm32 exti drivers.

Tested and validated against all STM32 series

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

### CODE REVIEW

File: drivers/interrupt_controller/exti_stm32.c [hunk 1]
```diff
@@ -117,88 +118,24 @@ void stm32_exti_enable(int line)
...
-			break;
-		}
-	}
-#elif CONFIG_SOC_SERIES_STM32F7X
-	if (line >= 5 && line <= 9) {
-		irqnum = EXTI9_5_IRQn;
-	} else if (line >= 10 && line <= 15) {
-		irqnum = EXTI15_10_IRQn;
-	} else if (line >= 0 && line <= 4) {
+	} else if {
```
Comment (tagunil) [inline drivers/interrupt_controller/exti_stm32.c]: Looks really weird!

File: drivers/interrupt_controller/exti_stm32.c [hunk 2]
```diff
@@ -625,6 +466,9 @@ static void __stm32_exti_connect_irqs(struct device *dev)
 		CONFIG_EXTI_STM32_EXTI15_10_IRQ_PRI,
 		__stm32_exti_isr_15_10, DEVICE_GET(exti_stm32),
 		0);
+#elif defined (ONFIG_SOC_SERIES_STM32F2X) || \
```
Comment (tagunil) [inline drivers/interrupt_controller/exti_stm32.c]: CONFIG?

Comment (erwango) [reply drivers/interrupt_controller/exti_stm32.c]: Oups, forgot to amend before pushing... Fixed, thanks!

File: drivers/interrupt_controller/exti_stm32.c (hunk 1 above)
Comment (erwango) [reply drivers/interrupt_controller/exti_stm32.c]: Yes, I missed some code in the pushed commit, fixed Txs

**Comment by erwango** [comment]
@tagunil , txs for review, this should be fixed now

**Comment by erwango** [comment]
recheck

**Comment by erwango** [comment]
@tagunil , can you review again?

**Comment by tagunil** [comment]
@erwango, it may be outside the scope of this cleanup PR, but why we still treat F2/F4/F7 differently by enabling IRQs for lines >=16 on these series and not enabling on others? Which way it should be done actually?
Otherwise LGTM.

**Review by tagunil — APPROVED** [review]

**Comment by erwango** [comment]
@tagunil, this is right, support could be enhanced, but this was not the aim of this PR. We can log a ticket for that purpose.

**Comment by avisconti** [comment]
@erwango 
quickly tested with microphone and sensor samples for ArgonKey and seems good.

File: drivers/interrupt_controller/exti_stm32.c [hunk 3]
```diff
@@ -214,45 +151,16 @@ void stm32_exti_enable(int line)
...
+#endif
+#if defined(CONFIG_SOC_SERIES_STM32F7X)
 		case 23:
 			irqnum = LPTIM1_IRQn;
 			break;
+#endif
+		default:
+			return;
 		}
 	}
```
Comment (avisconti) [inline drivers/interrupt_controller/exti_stm32.c]: case F2/F4/F7 seems in line with previous code.
But I'm not sure about F1/F3. In these case if line is above 15 what happens?

**Review by avisconti — COMMENTED** [review]
It is complex...

File: drivers/interrupt_controller/exti_stm32.c (hunk 3 above)
Comment (erwango) [reply drivers/interrupt_controller/exti_stm32.c]: @avisconti, I reworked a bit the change to be closer to previous code for L4.
For F1/F3, line > 15, code was obviously wrong, also we missed handling for not supported lines for F2/F4/F7 so reporting an error now.

Comment (avisconti) [reply drivers/interrupt_controller/exti_stm32.c]: > For F1/F3, line > 15, code was obviously wrong, also we missed handling for not supported lines for F2/F4/F7 so reporting an error now.

OK, got it now.

**Comment by avisconti** [comment]
I also re-tested the two commits on ArgonKey (microphone and sensor samples) and on 
disco_l475_iot1 (with lsm6dsl + trigger sample). It works fine.

File: drivers/interrupt_controller/exti_stm32.c [hunk 4]
```diff
@@ -1,21 +1,19 @@
...
- * Based on reference manuals:
- *   RM0008 Reference Manual: STM32F101xx, STM32F102xx, STM32F103xx, STM32F105xx
- *   and STM32F107xx advanced ARM-based 32-bit MCUs
- * and
- *   RM0368 Reference manual STM32F401xB/C and STM32F401xD/E
- *   advanced ARM-based 32-bit MCUs
- *
- * Chapter 10.2: External interrupt/event controller (EXTI)
+ * Driver is currently implemented to support following EXTI lines
+ * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 not supported
```
Comment (avisconti) [inline drivers/interrupt_controller/exti_stm32.c]: This is F1/F3/L4. Correct?

File: drivers/interrupt_controller/exti_stm32.c [hunk 5]
```diff
@@ -1,21 +1,19 @@
...
+ * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ
```
Comment (avisconti) [inline drivers/interrupt_controller/exti_stm32.c]: This is F0/L0. Correct?

Comment (erwango) [reply drivers/interrupt_controller/exti_stm32.c]: Actually this should be 
 * STM32F1/STM32F3: Lines 0 to 15. Lines > 15 not supported
 * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ.
Fixed, txs!

**Review by avisconti — APPROVED** [review]
OK, now seems fine. Thx for the comment at the top of the file. It helps.
I have also re-tested once more on ArgonKey and it works fine.

**Pull request merged**